Source/WebCore/page/PointerLockController.cpp

```diff
@@ -40,44 +42,91 @@
+    }
+};
+
 PointerLockController::PointerLockController(Page& page)
     : m_page(page)
 {
 }
 
-void PointerLockController::requestPointerLock(Element* target)
+PointerLockController::~PointerLockController()
+{
+}
+
+void PointerLockController::requestPointerLock(Element* target, std::optional<PointerLockOptions>&& options, RefPtr<DeferredPromise> promise)
 {
     if (!target || !target->isConnected() || m_documentOfRemovedElementWhileWaitingForUnlock) {
         enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        if (promise)
+            promise->reject(WrongDocumentError);
         return;
     }
 
     if (m_documentAllowedToRelockWithoutUserGesture != &target->document() && !UserGestureIndicator::processingUserGesture()) {
         enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        // If the request was not started from an engagement gesture and the Document has not previously released a successful Pointer Lock with exitPointerLock():
+        if (promise)
+            promise->reject(NotAllowedError);
         return;
     }
 
     if (target->document().isSandboxed(SandboxPointerLock)) {
         // FIXME: This message should be moved off the console once a solution to https://bugs.webkit.org/show_bug.cgi?id=103274 exists.
         target->document().addConsoleMessage(MessageSource::Security, MessageLevel::Error, "Blocked pointer lock on an element because the element's frame is sandboxed and the 'allow-pointer-lock' permission is not set."_s);
         enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        // If this's node document's active sandboxing flag set has the sandboxed pointer lock browsing context flag set:
+        if (promise)
+            promise->reject(SecurityError);
+        return;
+    }
+
+    if (options && options->unadjustedMovement && !m_page.chrome().client().pointerLockSupportsUnadjustedMovement()) {
+        // If options["unadjustedMovement"] is true and the platform does not support unadjustedMovement:
+        enqueueEvent(eventNames().pointerlockerrorEvent, target);
+        if (promise)
+            promise->reject(NotSupportedError);
         return;
     }
 
     if (m_element) {
         if (&m_element->document() != &target->document()) {
             enqueueEvent(eventNames().pointerlockerrorEvent, target);
+            // If the user agent's pointer-lock target is an element whose shadow-including root is not equal to this's shadow-including root, then:
+            if (promise)
+                promise->reject(InvalidStateError);
             return;
         }
         m_element = target;
-        enqueueEvent(eventNames().pointerlockchangeEvent, target);
-        m_page.pointerCaptureController().pointerLockWasApplied();
+        m_options = WTFMove(options);
+        if (m_lockPending) {
+            if (promise)
+                m_promises.append(promise.releaseNonNull());
```

Comment:
Maybe?: 

```suggestion
        if (m_lockPending && promise) {
              m_promises.append(promise.releaseNonNull());
```

Reply:
I can't make this change as-is. This isn't logically equivalent to the code I have. The else branch can only be taken if `!m_lockPending`. Whether the promise exists is just a function of feature flag `pointerLockOptionsEnabled`.

But that distinction is obviously too subtle. I will add a comments to make it more clear why this simplification cannot be applied for future readers of the code.